frc53_nft/src/state.rs

```diff
@@ -553,18 +585,134 @@ impl NFTState {
+                Ok(())
+            })?;
+
+        let next_cursor = next_key.map(|key| AmtCursor::new(self.token_data, key));
+        Ok((token_ids, next_cursor))
+    }
+
+    /// List the tokens owned by an actor. In the reference implementation this is may be a
+    /// prohibitievly expensive operation as it involves iterating over the entire token set.
+    /// Returns a bitfield of the tokens owned by the actor and a cursor to the next page of data.
+    pub fn list_owned_tokens<BS: Blockstore>(
+        &self,
+        bs: &BS,
+        owner: ActorID,
+        range_start: Option<AmtCursor>,
+        limit: Option<u64>,
+    ) -> Result<(TokenSet, Option<AmtCursor>)> {
+        let token_data_array = match &range_start {
+            Some(cursor) => self.get_token_amt_at_cursor(bs, cursor)?,
+            None => self.get_token_data_amt(bs)?,
+        };
+
+        // Build the TokenSet
+        let mut token_ids = TokenSet::new();
+        let (_, next_key) =
+            token_data_array.for_each_ranged(range_start.map(|r| r.index), limit, |i, data| {
+                if data.owner == owner {
+                    token_ids.set(i);
+                }
+                Ok(())
+            })?;
+
+        let next_cursor = next_key.map(|key| AmtCursor::new(self.token_data, key));
+        Ok((token_ids, next_cursor))
+    }
+
+    /// List all the token operators for a given token_id
+    pub fn list_token_operators<BS: Blockstore>(
+        &self,
+        bs: &BS,
+        token_id: TokenID,
+    ) -> Result<Vec<ActorID>> {
+        let token_data_array = self.get_token_data_amt(bs)?;
+        let token_data =
+            token_data_array.get(token_id)?.ok_or(StateError::TokenNotFound(token_id))?;
+
+        let operators = token_data.operators.iter().collect::<Vec<TokenID>>();
+        Ok(operators)
+    }
+
+    /// Enumerates tokens for which an account is an operator for a given owner
+    pub fn list_operator_tokens<BS: Blockstore>(
+        &self,
+        bs: &BS,
+        owner: ActorID,
+        operator: ActorID,
+        cursor: Option<AmtCursor>,
+        max: Option<u64>,
+    ) -> Result<ListOperatorTokensReturn> {
+        // Iterate over the owner's tokens and find those where the operator is approved
+        let mut operatable_tokens = TokenSet::new();
+        let token_array = self.get_token_data_amt(bs)?;
+        let (tokens, next_cursor) = self.list_owned_tokens(bs, owner, cursor, max)?;
+
+        // This step filters the returned TokenSet so the operatable_tokens may have size < max
+        tokens.iter().try_for_each(|token_id| -> Result<()> {
+            let token_data = token_array.get(token_id)?.ok_or_else(|| {
+                StateError::InvariantFailed(format!(
+                    "token {token_id} owned by {owner} not found in token amt",
+                ))
+            })?;
+
+            if token_data.operators.get(operator) {
+                operatable_tokens.set(token_id);
+            }
```

Comment:
If we remove the `owner` parameter here I expect this will simplify down to look a lot like list_owned_tokens.